tests/Unit/ParallelAlgorithms/LinearSolver/Schwarz/Test_Weighting.cpp

```diff
@@ -0,0 +1,136 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#include "Framework/TestingFramework.hpp"
+
+#include "DataStructures/DataVector.hpp"
+#include "DataStructures/Tensor/Tensor.hpp"
+#include "Domain/LogicalCoordinates.hpp"
+#include "Domain/Structure/Element.hpp"
+#include "Domain/Structure/Neighbors.hpp"
+#include "Domain/Structure/Side.hpp"
+#include "NumericalAlgorithms/Spectral/Mesh.hpp"
+#include "NumericalAlgorithms/Spectral/Spectral.hpp"
+#include "ParallelAlgorithms/LinearSolver/Schwarz/OverlapHelpers.hpp"
+#include "ParallelAlgorithms/LinearSolver/Schwarz/Weighting.hpp"
+
+namespace LinearSolver::Schwarz {
+
+namespace {
+template <size_t Dim>
+void test_weights_conservation(const Element<Dim>& element,
+                               const Mesh<Dim>& mesh,
+                               const size_t max_overlap) noexcept {
+  INFO("Weight conservation");
+  CAPTURE(Dim);
+  CAPTURE(element);
+  CAPTURE(mesh);
+  CAPTURE(max_overlap);
+  const auto logical_coords = logical_coordinates(mesh);
+  CAPTURE(logical_coords);
+  const size_t num_points = mesh.number_of_grid_points();
+  std::array<double, Dim> overlap_widths{};
+  for (size_t d = 0; d < Dim; ++d) {
+    const auto overlap_extent =
+        LinearSolver::Schwarz::overlap_extent(mesh.extents(d), max_overlap);
+    const auto collocation_points =
+        Spectral::collocation_points(mesh.slice_through(d));
+    gsl::at(overlap_widths, d) = LinearSolver::Schwarz::overlap_width(
+        overlap_extent, collocation_points);
+  }
+  CAPTURE(overlap_widths);
+  // The weights for the element data and those for data on intruding overlaps
+  // should sum to one.
+  const auto element_weights = LinearSolver::Schwarz::element_weight(
+      logical_coords, overlap_widths, element.external_boundaries());
+  CAPTURE(element_weights);
+  auto summed_weights = element_weights;
+  for (const auto& direction_and_neighbors : element.neighbors()) {
+    const auto& direction = direction_and_neighbors.first;
+    const size_t num_neighbors = direction_and_neighbors.second.size();
+    get(summed_weights) +=
+        num_neighbors * get(LinearSolver::Schwarz::intruding_weight(
+                            logical_coords, direction, overlap_widths,
+                            num_neighbors, element.external_boundaries()));
+  }
+  CHECK_ITERABLE_APPROX(get(summed_weights), DataVector(num_points, 1.));
```

Comment:
Is it true that if the weights sum to 1 that all of the corner logic has been correctly applied? If it's possible to mess up the corners and still get 1, then it's probably wise to have some simple 2 or 3 D case worked explicitly; e.g. a 3x3x3 element with a convenient choice for overlap width.

Reply:
I think so, yes. I test the weights on an element with all possible configurations of neighbors, so the conservation test is pretty thorough. However, looking at the tests again I found one possible failure mode that I added another test for: The exact shape of the weighting function wasn't fixed by the tests beyond requiring it goes from 0 to 1 and is 0.5 at the center, so I now test it explicitly by checking the weighting function shape against Eqn. (41) in https://arxiv.org/abs/1907.01572.